core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java

```diff
@@ -181,15 +182,21 @@ public RunT loadBuild(File dir) throws IOException {
     public final synchronized RunT newBuild() throws IOException {
         try {
             RunT lastBuild = getBuildClass().getConstructor(asJob().getClass()).newInstance(asJob());
+            var rootDir = lastBuild.getRootDir().toPath();
+            if (Files.isDirectory(rootDir)) {
+               LOGGER.warning(() -> "JENKINS-23152: " + rootDir + " already existed; will not overwrite with " + lastBuild + " but will create a fresh build #" + asJob().getNextBuildNumber());
+               return newBuild();
+            }
             builds.put(lastBuild);
             lastBuild.getPreviousBuild(); // JENKINS-20662: create connection to previous build
             return lastBuild;
         } catch (InvocationTargetException e) {
             LOGGER.log(Level.WARNING, String.format("A new build could not be created in job %s", asJob().getFullName()), e);
             throw handleInvocationTargetException(e);
-        } catch (ReflectiveOperationException | IllegalStateException e) {
-            LOGGER.log(Level.WARNING, String.format("A new build could not be created in job %s", asJob().getFullName()), e);
-            throw new LinkageError(e.getMessage(), e);
+        } catch (ReflectiveOperationException e) {
+            throw new LinkageError("A new build could not be created in " + asJob().getFullName() + ": " + e, e);
+        } catch (IllegalStateException e) {
```

Comment:
Only tangentially related, but note the inappropriate use of `LinkageError` in the PR description. This seems to be because #2125 added a multicatch which #5483 translated incorrectly.